scripts/lint-messages.js

```diff
@@ -0,0 +1,154 @@
+  let curClass
+  let lineIdx = 0
+  contents.forEach(line => {
+    lineIdx++
+
+    // find a MessageComponent class
+    const classMatch = classRegex.exec(line)
+    if (classMatch) {
+      curClass = classMatch[1]
+      return
+    }
+
+    // look for a message
+    if (curClass) {
+      const messageIndex = line.indexOf('this.messages(')
+      if (messageIndex > -1) {
+        // determine what the message path is
+        // if the first character is not ' or ` it's a variable
+        const firstChar = line[messageIndex + 14]
+        if (firstChar === "'") {
+          // isolate string so regex doesn't pick up anything before
+          const message = line.substring(
+            messageIndex + 15,
+            line.indexOf(firstChar, messageIndex + 16)
+          )
+          objectPath.set(foundMessages, `${curClass}.${message}`, true)
+        } else {
+          // TODO: analyze some kind of comment outlining message possibilities
+          console.warn(
+            `unable to determine message object path from variable or template string in file ${filePath} on line ${lineIdx}`
+          )
+        }
+      }
+    }
+  })
+}
+
+// iterate through files and tally up what was found and what wasn't
+exploreDir('lib')
+
+// analyze differences of messages in language files
+Object.keys(messages).forEach(langFile => {
+  const langMessages = messages[langFile]
+
+  function reconcileObject (obj, pathSoFar) {
+    Object.keys(obj).forEach(key => {
+      const nextPath = pathSoFar ? `${pathSoFar}.${key}` : key
+      if (typeof obj[key] === 'string') {
+        Object.keys(messages)
+          .filter(compareLangFile => compareLangFile !== langFile)
+          .forEach(compareLangFile => {
+            const compareLangMessages = messages[compareLangFile]['components']
+            if (obj[key] === objectPath.get(compareLangMessages, nextPath)) {
+              // FIXME: make this cause an error once we decide to have actual
+              // translations for all languages
+              console.warn(`WARNING: message ${nextPath} in ${langFile} is the same as the message in ${compareLangFile}`)
+            }
+          })
+      } else {
+        reconcileObject(obj[key], nextPath)
+      }
+    })
+  }
+
+  reconcileObject(langMessages['components'])
+})
+
+// analyze what defined messages in the language files were found and which were not
+Object.keys(messages).forEach(langFile => {
+  const langMessages = messages[langFile]
+
+  function reconcileObject (obj, pathSoFar) {
+    Object.keys(obj).forEach(key => {
+      const nextPath = pathSoFar ? `${pathSoFar}.${key}` : key
+      if (typeof obj[key] === 'string') {
+        if (!objectPath.get(foundMessages, nextPath)) {
+          // FIXME: make this cause an error once variables and
+          // template strings are accounted for
+          console.warn(`WARNING: message ${nextPath} in ${langFile} not found in any .js files`)
```

Comment:
How do we intend to account fully for variables? Is this really giving us valuable information if we're not accounting for variables and template strings? Maybe we should just report the number and location of messages defined with variables and template strings?

Reply:
We should be fully accounting for messages provided as variables and template strings.  I didn't get around to that bit yet.  I was thinking that we could add comments into the js files in a regular format that describe what possibilities of messages are available.  Then, the script would parse the comment.  Something like this:

```javascript
<ControlLabel>{this.messages(`${f}.label`)}</ControlLabel> // lint-message-possibilities: name.label, logoUrl.label
```